spark/v3.4/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceStaticInvoke.scala

```diff
@@ -0,0 +1,125 @@
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanRelation
+import org.apache.spark.sql.types.StructField
+import org.apache.spark.sql.types.StructType
+import scala.annotation.tailrec
+
+/**
+ * Spark analyzes the Iceberg system function to {@link StaticInvoke} which could not be pushed
+ * down to datasource. This rule will replace {@link StaticInvoke} to
+ * {@link ApplyFunctionExpression} for Iceberg system function in a filter condition.
+ */
+object ReplaceStaticInvoke extends Rule[LogicalPlan] with SQLConfHelper {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (systemFuncPushDownEnabled()) {
+      rewrite(plan)
+    } else {
+      plan
+    }
+  }
+
+  private def rewrite(plan: LogicalPlan): LogicalPlan = plan transform {
+    case filter @ Filter(condition, child) if icebergRelation(child) =>
+      val newCondition = replaceStaticInvoke(condition)
+      if (newCondition fastEquals condition) {
+        filter
+      } else {
+        filter.copy(condition = newCondition)
+      }
+  }
+
+  private def systemFuncPushDownEnabled(): Boolean = {
+    conf.getConfString(
+      SparkSQLProperties.SYSTEM_FUNC_PUSH_DOWN_ENABLED,
+      SparkSQLProperties.SYSTEM_FUNC_PUSH_DOWN_ENABLED_DEFAULT.toString
+    ).toBoolean
+  }
+
+  private def replaceStaticInvoke(expression: Expression): Expression = expression.transform {
+    case invoke: StaticInvoke =>
+      invoke.functionName match {
+        case "invoke" =>
+          if (invoke.arguments.forall(_.foldable)) {
+            // The invoke should be folded into constant
+            return invoke
+          }
+
+          // Adaptive from `resolveV2Function` in org.apache.spark.sql.catalyst.analysis.ResolveFunctions
+          val unbound = SparkFunctions.loadFunctionByClass(invoke.staticObject)
+          if (unbound == null) {
+            return invoke
+          }
+
+          val inputType = StructType(invoke.arguments.zipWithIndex.map {
+            case (exp, pos) => StructField(s"_$pos", exp.dataType, exp.nullable)
+          })
+
+          val bound = try {
+            unbound.bind(inputType)
+          } catch {
+            case _: Exception =>
+              return invoke
+          }
+
+          if (bound.inputTypes().length != invoke.arguments.length) {
+            return invoke
+          }
+
+          bound match {
+            case scalarFunc: ScalarFunction[_] =>
+              ApplyFunctionExpression(scalarFunc, invoke.arguments)
+            case _ => invoke
+          }
+      }
+  }
+
+  @tailrec
+  private def icebergRelation(plan: LogicalPlan): Boolean = {
+
+    def isIcebergTable(table: Table): Boolean = table match {
+      case _: SparkTable => true
+      case _ => false
+    }
+
+    plan match {
+      case relation: DataSourceV2Relation =>
+        isIcebergTable(relation.table)
+      case scanRelation: DataSourceV2ScanRelation =>
+        icebergRelation(scanRelation.relation)
```

Comment:
I don't think it would make sense to convert when there is a scan relation because that means the pushdown already occurred, right?

Reply:
Thanks for the correction. This is not needed anymore since we have relaxed the restriction.